### PR TITLE
8282026: Remove support for unbound sequence layouts

### PR DESCRIPTION
**Pull request by mcimadamore**
When doing a pass over the memory layout API, I couldn't help noting that the API implementation and specification are made more complex by the fact that we have to specify what happens when a sequence layout *without a size* is found. Support for unbounded sequence layout was added many moons ago, with the understanding that it would have made life for jextract easier.

This is not the case; all cases where jextract is using an unbounded sequence layout today can be replaced with using a simple zero-length sequence (which is allowed by the API, as are empty structs/unions).

I also found that the only real use case for using unbounded sequence layout came from a quirk in the API to obtain var handles from layout. That is, if we want to obtain an indexed var handle into a JAVA_INT, we have to do the following:

```
VarHandle intHandle = MemoryLayout.newSequenceLayout(JAVA_INT)
            .varHandle(PathElement.sequenceLayout());
```

In hindsight, this is just boilerplate: in this case the user only cares about the element layout (`JAVA_INT`), but the API wants the user to wrap that element layout in a new sequence layout, only to then provide a path inside the sequence layout (which will add a free dimension to the resulting var handle).

It is not hard to see that all the above can be simplified, by adding an extra API method on memory layouts:

```
VarHandle intHandle = JAVA_INT.arrayElementVarHandle(JAVA_INT);
```

This more explicit, succint, and provdies less opportunities for bugs to hide.

Therefore, this patch removes support for unbounded sequence element; the net effect is that now _all_ layouts have a size - which significantly simplifies the use sites where layout sizes are computed (no need to use optional there). All without losing much in terms of expressiveness: zero-length sequence layout can be used in basically the same way in which unbounded sequence layouts were used before.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282026](https://bugs.openjdk.java.net/browse/JDK-8282026): Remove support for unbound sequence layouts

### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/642/head:pull/642` \
`$ git checkout pull/642`

Update a local copy of the PR: \
`$ git checkout pull/642` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/642/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 642`

View PR using the GUI difftool: \
`$ git pr show -t 642`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/642.diff">https://git.openjdk.java.net/panama-foreign/pull/642.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/foreign/MemoryLayout.java [hunk 1]
```diff
@@ -481,6 +485,30 @@ default MethodHandle sliceHandle(PathElement... elements) {
...
 
+    /**
+     * Creates a <em>strided</em> method handle which, given a memory segment, returns a {@linkplain MemorySegment#asSlice(long,long) slice}
+     * corresponding to the layout selected by a given layout path, where the path is considered rooted in this layout.
+     * The returned method handle can effectively slice a given memory segment at multiple starting offsets, using a <em>dynamic</em> index
+     * (of type {@code long}), which is multiplied by this layout size and then added to the offset of the selected layout.
+     * Equivalent to the following code:
+     * {@snippet lang=java :
+     * MemoryLayout.sequenceLayout(Long.MAX_VALUE, this)
+     *             .varHandle(PathElement.sequenceElement());
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:496]:
```suggestion
     * MemoryLayout.sequenceLayout(Long.MAX_VALUE, this)
     *             .sliceHandle(PathElement.sequenceElement());
```
?

File: src/java.base/share/classes/java/lang/foreign/SequenceLayout.java [hunk 2]
```diff
@@ -34,9 +34,9 @@
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/lang/foreign/SequenceLayout.java:34]: Unused import `java.util.OptionalLong;`, suggest checking other files too.

**Review by PaulSandoz — COMMENTED** [review]
I don't see any tests for `arrayElementSliceHandle`.

**Comment by mcimadamore** [comment]
After some more thinking I decided to drop `arrayElementSliceHandle` for the time being. The "sliceHandle" method is already quite nichey, and you can always get there by wrapping into a sequence layout and then unwrap with a sequence element path, to get a strided MH.

In other words, I think the API should make it easy to express var handles, which are the most common form - other things should be possible as well (and they are) but not sure they are worth the API estate.

File: src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java [hunk 3]
```diff
@@ -242,7 +238,7 @@ public MemoryLayout map(UnaryOperator<MemoryLayout> op) {
     private MemoryLayout dup(MemoryLayout oldLayout, MemoryLayout newLayout) {
         newLayout = newLayout.withBitAlignment(oldLayout.bitAlignment());
         if (oldLayout.name().isPresent()) {
-            newLayout.withName(oldLayout.name().get());
+            newLayout = newLayout.withName(oldLayout.name().get());
```
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:241]: oof, nice catch!

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**